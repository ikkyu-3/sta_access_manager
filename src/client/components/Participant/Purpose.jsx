// @flow
import type { ParticipantPurposeProps } from "@types";

import * as React from "react";
import { MEET_UP, STUDY, WORK, CIRCLE } from "@constants";
import styles from "./style.css";

function Purpose(props: ParticipantPurposeProps): React.Node {
  const { purpose } = props;
  const { columnPurpose, border, work, study, meetUp, circle } = styles;

  let background: string = "";
  let value: string = "";

  if (purpose === MEET_UP) {
    background = meetUp;
    value = "勉強会";
  }
  if (purpose === STUDY) {
    background = study;
    value = "自習";
  }
  if (purpose === WORK) {
    background = work;
    value = "仕事";
  }
  if (purpose === CIRCLE) {
    background = circle;
    value = "サークル";
  }

  return (
    <div className={`${columnPurpose} ${border} ${background}`}>{value}</div>
  );
}
export default Purpose;
