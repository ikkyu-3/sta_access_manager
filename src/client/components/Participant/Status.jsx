// @flow
import type { ParticipantStatusProps } from "@types";

import * as React from "react";
import styles from "./style.css";

function Status(props: ParticipantStatusProps): React.Node {
  const { isEntry } = props;
  const { columnStatus, entry, exit } = styles;

  const className = isEntry
    ? `${columnStatus} ${entry}`
    : `${columnStatus} ${exit}`;
  const value = isEntry ? "入" : "退";

  return <div className={className}>{value}</div>;
}
export default Status;
