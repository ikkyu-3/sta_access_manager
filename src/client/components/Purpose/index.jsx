// @flow
import type { PurposeProps } from "@types";

import * as React from "react";
import styles from "./style.css";

class Purpose extends React.Component<PurposeProps> {
  componentDidMount() {
    this.props.checkId();
  }

  componentWillUnmount() {
    this.props.removeId();
  }

  render() {
    const { onPurposeButtonClick } = this.props;
    const { container, meetup, work, study, circle, message } = styles;

    return (
      <div className={container}>
        <button id="meetup" className={meetup} onClick={onPurposeButtonClick}>
          勉強会
        </button>
        <button id="work" className={work} onClick={onPurposeButtonClick}>
          仕事
        </button>
        <button id="study" className={study} onClick={onPurposeButtonClick}>
          自習
        </button>
        <button id="circle" className={circle} onClick={onPurposeButtonClick}>
          サークル
        </button>
        <div className={message}>目的を選択して下さい</div>
      </div>
    );
  }
}
export default Purpose;
