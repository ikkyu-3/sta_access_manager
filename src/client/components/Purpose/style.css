@charset "utf-8";
.container {
  display: grid;
  position: relative;
  grid-template-columns: 1fr 1fr;
  grid-template-rows: 1fr 1fr;
}

.container > button {
  width: 100%;
  color: gray;
  margin: 0;
  padding: 0;
  font-size: 28px;
  font-weight: bold;
  border: none;
  box-sizing: border-box;
  outline: none;
}

.meetup {
  border-right: 2px solid lightgray !important;
  border-bottom: 2px solid lightgray !important;
  border-radius: 5px 0 0 0;
  background-color: #f8f189;
}

.work {
  border-left: 2px solid lightgray !important;
  border-bottom: 2px solid lightgray !important;
  border-radius: 0 5px 0 0;
  background-color: #bbfeff;
}

.study {
  border-right: 2px solid lightgray !important;
  border-top: 2px solid lightgray !important;
  border-radius: 0 0 0 5px;
  background-color: #bcf3cb;
}

.circle {
  border-left: 2px solid lightgray !important;
  border-top: 2px solid lightgray !important;
  border-radius: 0 0 5px 0;
  background-color: #f4d8f5;
}

.message {
  display: flex;
  position: absolute;
  width: 250px;
  height: 60px;
  left: calc(50% - 125px);
  top: calc(50% - 30px);
  justify-content: center;
  align-items: center;
  color: gray;
  font-size: 20px;
  font-weight: bold;
  background-color: white;
  border: 4px solid lightgray;
  border-radius: 10px;
  box-sizing: border-box;
}
