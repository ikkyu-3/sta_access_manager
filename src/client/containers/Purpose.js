// @flow
import type { State, Dispatch } from "@types";

import { connect } from "react-redux";
import { push } from "react-router-redux";

import { requestStart, requestEnd, removeId } from "@actions";
import Purpose from "@components/Purpose";
import {
  API_POST_ENTRY,
  TOP_PATH,
  COMPLETION_ENTRANCE_PATH,
  ERROR_BAD_REQUEST_PATH,
  ERROR_INTERNAL_SERVER_ERROR_PAHT,
  MEET_UP,
  STUDY,
  WORK,
  CIRCLE,
  UNKNOWN
} from "@constants";
import createApiUrl from "@modules/createApiUrl";
import request from "@modules/request";

const mapStateToProps = (state: State) => ({ state });

const mapDispatchToProps = (dispatch: Dispatch) => ({
  dispatch,
  removeId: () => {
    dispatch(removeId());
  }
});

const getPurpose = (purpose: string): string => {
  switch (purpose) {
    case "meetup":
      return MEET_UP;
    case "study":
      return STUDY;
    case "work":
      return WORK;
    case "circle":
      return CIRCLE;
    default:
      return "";
  }
};

const mergeProps = (stateProps, dispatchProps, ownProps) => ({
  ...dispatchProps,
  ...ownProps,
  checkId: () => {
    if (!/^[0-9a-f]{16}$/.test(stateProps.state.id)) {
      dispatchProps.dispatch(removeId());
      dispatchProps.dispatch(push(TOP_PATH));
    }
  },
  onPurposeButtonClick: async (
    e: SyntheticEvent<HTMLButtonElement>
  ): Promise<void> => {
    if (!(e.target instanceof window.HTMLButtonElement)) return;

    const purpose = getPurpose(e.target.id);
    if (purpose === UNKNOWN) return;

    const { id } = stateProps.state;

    const { dispatch } = dispatchProps;
    const { method, path } = API_POST_ENTRY;
    const url = createApiUrl(path);
    const data = { id, purpose };

    // 通信処理
    try {
      dispatch(requestStart());
      await request({ url, method, data });
      dispatch(requestEnd());

      dispatch(push(COMPLETION_ENTRANCE_PATH));
    } catch (error) {
      dispatch(requestEnd());

      if (error.status === 500) {
        dispatch(push(ERROR_INTERNAL_SERVER_ERROR_PAHT));
      } else {
        dispatch(push(ERROR_BAD_REQUEST_PATH));
      }
    }
  }
});

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)(
  Purpose
);
