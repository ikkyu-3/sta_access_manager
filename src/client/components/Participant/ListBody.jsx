// @flow
import type { ParticipantListBodyProps, User } from "@types";

import * as React from "react";
import Purpose from "./Purpose";
import Status from "./Status";
import styles from "./style.css";

function ListBody(props: ParticipantListBodyProps): React.Node {
  const { body, columnName, border, buttom } = styles;
  const { listData } = props;

  const { length } = listData;
  const listItems = listData.map((data: User, index: number): React.Node => {
    const { id, user } = data;
    const { name, purpose, isEntry } = user;
    const listButtom: string = index === length - 1 && index > 5 ? "" : buttom;
    const key = `list-${id}`;

    return (
      <li key={key} className={listButtom}>
        <div className={`${columnName} ${border}`}>{name}</div>
        <Purpose purpose={purpose} />
        <Status isEntry={isEntry} />
      </li>
    );
  });

  return (
    <div className={body}>
      <ul>{listItems}</ul>
    </div>
  );
}
export default ListBody;
