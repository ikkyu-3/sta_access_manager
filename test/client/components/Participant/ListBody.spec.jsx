import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import ListBody from "@components/Participant/ListBody";
import Purpose from "@components/Participant/Purpose";
import Status from "@components/Participant/Status";
import { shortParticipantData, longParticipantData } from "../testData";

describe("Participant.ListBody.jsxのテスト", () => {
  describe("スナップショット", () => {
    it("正しいレンダリング", () => {
      const tree = renderer
        .create(<ListBody listData={shortParticipantData} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("コンポーネントのテスト", () => {
    const listItems = shallow(
      <ListBody listData={shortParticipantData} />
    ).find("li");

    it("各li要素には、listDataで渡されたnameの値が表示されている", () => {
      listItems.forEach((node, index) => {
        expect(node.find(".columnName").text()).toBe(
          shortParticipantData[index].name
        );
      });
    });

    it("Purposeコンポーネントのpurposeに、listDataで渡されたpurposeの値が渡されている", () => {
      listItems.forEach((node, index) =>
        expect(
          node.contains(
            <Purpose purpose={shortParticipantData[index].purpose} />
          )
        )
      );
    });

    it("StatusコンポーネントのisEntryに、listDataで渡されたisEntryの値が渡されている", () => {
      listItems.forEach((node, index) =>
        expect(
          node.contains(
            <Status isEntry={shortParticipantData[index].isEntry} />
          )
        )
      );
    });

    it("listDataの長さが5以下の時、全ての要素はcssクラス: buttomを持つ", () => {
      listItems.forEach(node => {
        expect(node.hasClass("buttom")).toBeTruthy();
      });
    });

    it("listDataの長さが6以上の時、最後の要素のみcssクラス: buttomを持たない", () => {
      const longListItems = shallow(
        <ListBody listData={longParticipantData} />
      ).find("li");
      const { length } = longListItems;

      longListItems.forEach((node, index) => {
        if (index === length - 1) {
          expect(node.hasClass("buttom")).toBeFalsy();
        } else {
          expect(node.hasClass("buttom")).toBeTruthy();
        }
      });
    });
  });
});
