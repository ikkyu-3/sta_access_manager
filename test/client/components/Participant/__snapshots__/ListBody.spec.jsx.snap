// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Participant.ListBody.jsxのテスト スナップショット 正しいレンダリング 1`] = `
<div
  className="body"
>
  <ul>
    <li
      className="buttom"
    >
      <div
        className="columnName border"
      >
        name1
      </div>
      <div
        className="columnPurpose border study"
      >
        自習
      </div>
      <div
        className="columnStatus entry"
      >
        入
      </div>
    </li>
    <li
      className="buttom"
    >
      <div
        className="columnName border"
      >
        name2
      </div>
      <div
        className="columnPurpose border study"
      >
        自習
      </div>
      <div
        className="columnStatus entry"
      >
        入
      </div>
    </li>
    <li
      className="buttom"
    >
      <div
        className="columnName border"
      >
        name3
      </div>
      <div
        className="columnPurpose border study"
      >
        自習
      </div>
      <div
        className="columnStatus entry"
      >
        入
      </div>
    </li>
    <li
      className="buttom"
    >
      <div
        className="columnName border"
      >
        name4
      </div>
      <div
        className="columnPurpose border study"
      >
        自習
      </div>
      <div
        className="columnStatus entry"
      >
        入
      </div>
    </li>
    <li
      className="buttom"
    >
      <div
        className="columnName border"
      >
        name5
      </div>
      <div
        className="columnPurpose border study"
      >
        自習
      </div>
      <div
        className="columnStatus exit"
      >
        退
      </div>
    </li>
  </ul>
</div>
`;
